packages/client/src/links/httpTuplesonLink.ts

```diff
@@ -0,0 +1,144 @@
+import { unstable_createTsonAsyncOptions } from '@trpc/server/shared';
+import { createTsonParseAsync, TsonAsyncOptions } from 'tupleson';
+import { NonEmptyArray } from '../internals/types';
+import { HTTPBatchLinkOptions } from './HTTPBatchLinkOptions';
+import {
+  createHTTPBatchLink,
+  RequesterFn,
+} from './internals/createHTTPBatchLink';
+import { getTextDecoder } from './internals/getTextDecoder';
+// Stream parsing adapted from https://www.loginradius.com/blog/engineering/guest-post/http-streaming-with-nodejs-and-fetch-api/
+import {
+  fetchHTTPResponse,
+  getBody,
+  getUrl,
+  HTTPBaseRequestOptions,
+} from './internals/httpUtils';
+import { TextDecoderEsque } from './internals/streamingUtils';
+import { HTTPHeaders, Operation } from './types';
+
+export interface HTTPTuplesonLinkOptions extends HTTPBatchLinkOptions {
+  /**
+   * Will default to the webAPI `TextDecoder`,
+   * but you can use this option if your client
+   * runtime doesn't provide it.
+   */
+  textDecoder?: TextDecoderEsque;
+  tuplesonOptions?: Partial<TsonAsyncOptions>;
+}
+
+async function* readableStreamToAsyncIterable<T>(
+  stream: ReadableStream<T>,
+): AsyncIterable<T> {
+  // Get a lock on the stream
+  const reader = stream.getReader();
+
+  try {
+    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+    while (true) {
+      // Read from the stream
+      const result = await reader.read();
+
+      // Exit if we're done
+      if (result.done) {
+        return;
+      }
+
+      // Else yield the chunk
+      yield result.value;
+    }
+  } finally {
+    reader.releaseLock();
+  }
+}
+
+async function* mapIterable<T, TValue>(
+  iterable: AsyncIterable<T>,
+  fn: (v: T) => TValue,
+): AsyncIterable<TValue> {
+  for await (const value of iterable) {
+    yield fn(value);
+  }
+}
+
+const streamingJsonHttpRequester = (
+  opts: HTTPBaseRequestOptions & {
+    headers: () => HTTPHeaders | Promise<HTTPHeaders>;
+    textDecoder: TextDecoderEsque;
+    parseAsync: ReturnType<typeof createTsonParseAsync>;
+  },
+) => {
+  const ac = opts.AbortController ? new opts.AbortController() : null;
+  const responsePromise = fetchHTTPResponse(
+    {
+      ...opts,
+      contentTypeHeader: 'application/json',
+      batchModeHeader: 'stream',
+      getUrl,
+      getBody,
+    },
+    ac,
+  );
+  const cancel = () => ac?.abort();
+  const textDecoder = new TextDecoder();
+  const promise = responsePromise.then(async (res) => {
+    const stringIterator = mapIterable(
+      readableStreamToAsyncIterable(res.body as any),
```

Comment:
Not sure what is different from our tests in tupleson, but `res.body` doesn't have `.getReader()`


@Sheraff - any idea?

Reply:
It depends whether it's a response from Web API or a response from Node IIRC. It's been a while so I don't really remember but i had left these comments in https://github.com/trpc/trpc/blob/main/packages/client/src/links/internals/parseJSONStream.ts

```ts
  /**
   * As given by `(await fetch(url)).body`
   */
  readableStream: NodeJS.ReadableStream | WebReadableStreamEsque;
```

```ts
  // we handle 2 different types of streams, this if where we figure out which one we have
  if ('getReader' in readableStream) {
    await readStandardChunks(readableStream, onChunk);
  } else {
    await readNodeChunks(readableStream, onChunk);
  }
```

```ts
/**
 * Handle NodeJS stream
 */
function readNodeChunks(
  stream: NodeJS.ReadableStream,
  onChunk: (chunk: Uint8Array) => void,
) {
  return new Promise<void>((resolve) => {
    stream.on('data', onChunk);
    stream.on('end', resolve);
  });
}

/**
 * Handle WebAPI stream
 */
async function readStandardChunks(
  stream: WebReadableStreamEsque,
  onChunk: (chunk: Uint8Array) => void,
) {
  const reader = stream.getReader();
  let readResult = await reader.read();
  while (!readResult.done) {
    onChunk(readResult.value);
    readResult = await reader.read();
  }
}
```

Reply:
Nice, the nodejs one is an async iterable so i could just do `for await (....)`